toolkit/tools/internal/retry/retry_test.go

```diff
@@ -0,0 +1,133 @@
+package retry
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var defaultTestTime = time.Millisecond * 100
+
+// Tests won't run with perfect timing, so we need to allow for some fudge factor when we check maximum duration.
+var timingFudgeFactor = time.Millisecond * 100
+
+func TestZeroBase(t *testing.T) {
+	// Two failures, 0.0 multiplier, should be 0 second delay
+	base := 0.0
+	fails := 2
+	val := calculateExpDelay(fails, defaultTestTime, base)
+	assert.Equal(t, time.Duration(0), val)
+}
+
+func TestZeroI(t *testing.T) {
+	// Zero failures, 1.0 base, should be 0 second delay
+	fails := 0
+	base := 1.0
+	val := calculateExpDelay(fails, defaultTestTime, base)
+	assert.Equal(t, time.Second*0, val)
+
+	// Three failures, 1.0 base, should still be 0 second delay
+	fails = 3
+	base = 0.0
+	val = calculateExpDelay(fails, defaultTestTime, base)
+	assert.Equal(t, time.Second*0, val)
```

Comment:
Ideally, I'd try to test only one scenario per test, so that it also can fail only for one reason. Bundling multiple scenarios into one test makes it take longer to find the one, which is failing.